src/commands/keygen.rs

```diff
@@ -0,0 +1,380 @@
+            Some(SecAlg::RSASHA256) => GenerateParams::RsaSha256 { bits },
+            Some(SecAlg::ECDSAP256SHA256) => GenerateParams::EcdsaP256Sha256,
+            Some(SecAlg::ECDSAP384SHA384) => GenerateParams::EcdsaP384Sha384,
+            Some(SecAlg::ED25519) => GenerateParams::Ed25519,
+            Some(SecAlg::ED448) => GenerateParams::Ed448,
+            Some(_) => unreachable!(),
+            None => {
+                return Err("Missing algorithm (-a) option".into());
+            }
+        };
+
+        let Some(name) = name else {
+            return Err("Missing domain name argument".into());
+        };
+
+        Ok(Self {
+            algorithm,
+            make_ksk,
+            #[cfg(target_family = "unix")]
+            create_symlinks,
+            #[cfg(target_family = "unix")]
+            force_symlinks,
+            name,
+        })
+    }
+}
+
+impl From<Keygen> for Command {
+    fn from(value: Keygen) -> Self {
+        Self::Keygen(value)
+    }
+}
+
+impl Keygen {
+    fn parse_algorithm(value: &str) -> Result<GenerateParams, clap::Error> {
+        match value {
+            "RSASHA256" => return Ok(GenerateParams::RsaSha256 { bits: 2048 }),
+            "ECDSAP256SHA256" => return Ok(GenerateParams::EcdsaP256Sha256),
+            "ECDSAP384SHA384" => return Ok(GenerateParams::EcdsaP384Sha384),
+            "ED25519" => return Ok(GenerateParams::Ed25519),
+            "ED448" => return Ok(GenerateParams::Ed448),
+            _ => {}
+        }
+
+        // TODO: Remove attrs when more RSA algorithms are added.
+        #[allow(clippy::collapsible_match)]
+        if let Some((name, params)) = value.split_once(':') {
+            #[allow(clippy::single_match)]
+            match name {
+                "RSASHA256" => {
+                    let bits: u32 = params.parse().map_err(|err| {
+                        clap::Error::raw(
+                            clap::error::ErrorKind::InvalidValue,
+                            format!("invalid RSA key size '{params}': {err}"),
+                        )
+                    })?;
+                    return Ok(GenerateParams::RsaSha256 { bits });
+                }
+                _ => {}
+            }
+        }
+
+        Err(clap::Error::raw(
+            clap::error::ErrorKind::InvalidValue,
+            format!("unrecognized algorithm '{value}'"),
+        ))
+    }
+
+    pub fn execute(self, env: impl Env) -> Result<(), Error> {
+        let mut stdout = env.stdout();
+
+        let params = self.algorithm;
+
+        // The digest algorithm is selected based on the key algorithm.
+        let digest_alg = match params.algorithm() {
+            SecAlg::RSASHA256 => DigestAlg::SHA256,
+            SecAlg::ECDSAP256SHA256 => DigestAlg::SHA256,
+            SecAlg::ECDSAP384SHA384 => DigestAlg::SHA384,
+            SecAlg::ED25519 => DigestAlg::SHA256,
+            SecAlg::ED448 => DigestAlg::SHA256,
+            _ => unreachable!(),
```

Comment:
You might be right, but I wonder if `SecAlg` makes sense here if we always ignore some variants? Shouldn't we then take `DigestAlg`?

Reply:
Hm.  We could add an option to explicitly specify a digest, but these defaults are the same that ldns uses.  I suspect that using an irregular signature/digest algorithm combination is not a good idea.

Reply:
Oh yeah I missed that it's not 1 to 1.